### DISCUSSION TITLE
plotting halo merger trees

### DISCUSSION THREAD

**Discussion by Luke-Williams-01** [galacticusorg/galacticus#741] [Q&A] [answered]
Hello!

This may be a trivial question, but I am trying to plot a halo merger tree using the output of galacticus. Two things I've been using are the tutorial on generating galaxy merger trees:

https://github.com/galacticusorg/galacticus/wiki/Tutorial:-Generating-galaxy-merger-trees

and the output from this I ran through this python file:

https://github.com/galacticusorg/galacticus/blob/master/scripts/analysis/galaxyMergerTree.py

This plotted the graph seen at the end of the Galaxy merger trees just fine. I assumed it would be relatively trivial to swap out the property names at the start of the python file for the corresponding halo related terms, e.g "galaxyMergerTreeNodeIndex" to "nodeIndex", "galaxyMergerTreeMassStellarTotal" to "basicMass" etc.

However, this didn't seem to work and also I don't believe I am substituting the correct terms to make the plot I want. I tried for example, looking through the Galacticus Development pdf, and found the "massHalo" to substitute for the stellar mass, however, I think I added it to the xml file incorrectly.

I would very much appreciate any assistance you could provide on whether the approach I using is useful, or indeed if there is some far simpler way to create a merger tree plot which is only concerned with showing the merger history of the dark matter halos.

Many thanks,

Luke Williams

**Comment by abensonca** [comment, marked as answer]
Hi Luke: 

Unfortunately it's not that simple! There's a fundamental difference here between the "galaxy merger trees" that are plotted by this script, and a "halo merger tree". The reason for this is that "halo merging" (in Galacticus' definition) means when two halos merge and the smaller one becomes a subhalo in the larger one (but still exists as a distinct object). "Galaxy merging" means when that subhalo (and the galaxy it contains) merges with the central halo (and the galaxies of the two halos merge) - which happens after (often much after) the halo merging.

So, probably these galaxy merger trees are just not what you want if your goal is to plot halo merger trees.

Two options that might work for you:

1. Make use of the `dumpToGraphViz` merger tree operator by adding:
```
<mergerTreeOperator value="dumpToGraphViz"/>
```
to your parameter file. See [here](https://github.com/galacticusorg/galacticus/releases/download/bleeding-edge/Galacticus_Physics.pdf#subsection.3.99.6) for details on the options to this operator. This will output tree data to a file that can be run through the [`dot`](https://graphviz.org/doc/info/lang.html) tool to plot tree diagrams.

2. You could use the `export` merger tree operator by adding:
```
<mergerTreeOperator value="export">
```
to your parameter file. See [here](https://github.com/galacticusorg/galacticus/releases/download/bleeding-edge/Galacticus_Physics.pdf#subsection.3.99.7) for options.

This will output the merger tree structure to a file `'galacticusExportedTrees.hdf5` by default, in the format specified [here](https://github.com/galacticusorg/galacticus/wiki/Merger-Tree-File-Format). You could then read this data and plot it using a Python script (probably you could re-use of some of the code in the Python script used to plot galaxy merger trees). This would give you more control over how the tree is plotted, but would take more work.

**Reply by Luke-Williams-01** [reply]
Hi Andrew,

Thank you for your quick response! I thought it might not be as simple as I first thought, I'll mark this question as answered and will likely try both methods you have suggested over the next few days.

I also had another question more generally about the way the parameter files are constructed - I noticed when reading through the Galacticus Development and Galacticus Physics documentation, that various properties can be written to the output file using the "nodePropertyExtractor". I'm still quite new to using Galacticus and was wondering how generally I include these in my parameter files. Many of the "nodePropertyExtractor" in the Development document give a parent class and in the Physics document sometimes underneath the definition there will be words to the effect of "classes used" or "implemented by".

I was mainly wondering if, when I think about using the "nodePropertyExtractor", placing the parent class first is the right way to go about things?

Thank you for your time,
Luke

**Comment by abensonca** [comment]
That's right - the `nodePropertyExtractor` class allows you to control what properties are written to the output file. More precisely, if you include the `<mergerTreeOutputter value="standard">` parameter - which you almost certainly do - it will look for any `nodePropertyExtractor` in your parameter file and use that to extract properties from each "node" (i.e. halo) and write them to the output file.

Typically you don't need to be concerned about the parent classes of any of the `nodePropertyExtractor`s (unless you're creating a new member of the class). Instead, you would include something like this in your parameter file:
```
      <nodePropertyExtractor value="multi">
        <nodePropertyExtractor value="time"/>
        <nodePropertyExtractor value="virialProperties"/>
        <nodePropertyExtractor value="nodeIndices"/>
      </nodePropertyExtractor>
```
In this case, you're asking it to output the time (i.e. the time at which each halo exists), the virial properties (virial radius, velocity, and temperature), and the node indices (which include things like the node index, the index of its host halo, etc.). These are wrapped inside the `<nodePropertyExtractor value="multi">` which combines all of these into a single effective extractor which will then be used by the outputter. You can add as many different extractor in this list as you want.

**Reply by Luke-Williams-01** [reply]
Thank you! I suspected that the "multi" extractor was a catch all that I could just add to but I was a little unsure. Many thanks
